issue-992: pySCG-0042: make code examples clearly demonstrate uninten… (#1108)
ossf/wg-best-practices-os-developers: docs/Secure-Coding-Guide-for-Python/01_introduction/pyscg-0042/README.md

_[noncompliant01.py](noncompliant01.py):_

```python
# SPDX-FileCopyrightText: OpenSSF project contributors
# SPDX-License-Identifier: MIT
"""Non-Compliant Code Example"""

ARG_MAX_BYTES = 16  # small buffer
argbuf = bytearray(ARG_MAX_BYTES)

USED = 10                       # 10 bytes already used
ARG = b"AAAAAAAA"               # attacker-controlled, needs 8 + 1 = 9 bytes
LENGTH = len(ARG) + 1           # 9 (include NUL terminator)

REMAINING = ARG_MAX_BYTES - USED + LENGTH  # Wrong: 15, expected -3

if LENGTH > REMAINING:
    print("Rejected: not enough space")
else:
    print(f"Bounds check passed: remaining={REMAINING}")

```

_Example output of `noncompliant01.py`:_

```bash
Bounds check passed: remaining=15
```

The bounds check incorrectly passes because the expression evaluates left-to-right as `(16 - 10) + 9 = 15` instead of the intended `16 - (10 + 9) = -3`. This allows the attacker-controlled data to overflow the buffer.

## Compliant Solution

> The compliant solution fixes the operator precedence by comparing additions instead of subtracting, making the code clear, with no precedence trap. While adding parentheses like `ARG_MAX_BYTES - (USED + LENGTH)` would fix the precedence issue also, restructuring to `USED + LENGTH > ARG_MAX_BYTES` is clearer and avoids confusing negative values.

Reviewer: Nitpick: I found the wording a bit confusing. How about this?

```suggestion
The compliant solution avoids the operator precedence trap by comparing additions instead of subtracting, making the code clearer. While adding parentheses like `ARG_MAX_BYTES - (USED + LENGTH)` would also fix the precedence issue, restructuring to `USED + LENGTH > ARG_MAX_BYTES` is more readable and avoids confusion caused by negative values.
```
